Cancel queued activities if AbortOnResupply is set.

pchote opened the pull request:
This PR aims to fix #17525 by recreating the behaviour from `release-20190314` for both regular attacks and attack moves. This is a hack, but IMO not significantly worse than the rest of `AbortOnResupply`.

This is simple enough to also pick to the hotfix, if not rejected for being awful.

Fixes #17525.
Fixes #17249.

tovl commented:
This looks like it would trigger the bug where activities go directly from queued to canceling. That would need to be fixed first, which would put this beyond the scope of a hotfix, I think.

Perhaps a less hacky solution would be to always return false after queueing `ReturnToBase` and then calling `Cancel(self)` on the next tick (i.e. if `returnToBase == true && attackAircraft.Info.AbortOnResupply`).

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -120,6 +120,12 @@ public override bool Tick(Actor self)
+				// HACK: The gameplay expectation for AbortOnResupply is to cancel all queued activities, but only
+				// if the attack was explicitly ordered (i.e. FlyAttack is a top-level activity).
+				// Implicit orders (i.e. FlyAttack is a child activity) should not be disrupted
+				if (attackAircraft.Info.AbortOnResupply && NextActivity != null && self.CurrentActivity == this)
```
The `self.CurrentActivity` check doesn't work if an attack order is queued after another activity while having no ammo. CurrentActivity isn't updated until the end of `RunActivity`, so this check will always be false for the first tick.

tovl replied:
The only time `FlyAttack` is called as a child activity, is when it is called by `AttackMoveActivity`. So it would be cleaner IMO to explicitly define this behaviour there, by passing a boolean argument down to the childactivity. This avoids the issue with checking `self.CurrentActivity == this` entirely, which is a hack in itself.

pchote replied:
I considered that, but the problem is that `AttackMoveActivity` works through an interface, and exposing this workaround to all attack traits is IMO a bigger hack than the (now updated) check here.

If this still crosses our collective hack tolerance threshold I would rather abandon the idea completely.

pchote commented:
> Perhaps a less hacky solution would be...

It is slightly more complicated than this because we also need to suppress the target lines for the later activities. This will need some more work to solve the `CurrentActivity` issue that @abcdefg30 found on IRC - if there isn't a simple solution that fixes both issues then we can go back to the drawing board for next + 1.

pchote commented:
> This looks like it would trigger the bug where activities go directly from queued to canceling.

A possible work around could be to call `Cancel()` on the `FlyActivity`, which will then null out `NextActivity` and avoid queued cancellation bug. We then queue the RTB activity as a standard activity instead of a child.

This would be an even worse hack, and I don't like the idea of allowing activities to null NextActivity directly, so I will look for a cleaner solution.

pchote commented:
Updated. This should fix the issues above, but means that the RTB target line is no longer shown. I'm not sure whether this could be considered an acceptable tradeoff?

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -120,6 +120,19 @@ public override bool Tick(Actor self)
+					Cancel(self);
+				}
+
 				QueueChild(new ReturnToBase(self));
```
Queuing this as a top level activity in the branch above should fix the target line issue.

pchote commented:
I don't really understand your argument to be honest - there is indeed no specific expectation on that specific named flag, but there is an expectation in the way that units behave, which is what that flag controls. The activity rework made an unexpected change to these behaviours, which this PR restores.

pchote committed:
Add AttackSource enum.

tovl commented:
> The activity rework made an unexpected change to these behaviours, and this PR aims to correct that.

It wasn't unexpected or unintentional though (at least not to me). Also, keep in mind that the behaviour on the previous release was exactly the same as on the current release ... if the queued order was given while the aircraft was on it's way back to base. So this isn't actually recreating the behaviour from previous release.

It's also pretty bad UX to let the player queue orders (and provide visual feedback for them) only to unpredictably drop them at some point halfway through. The simple fact that you can queue orders at all makes the `AbortOnResupply` behaviour completely counter to naive expectations. Indeed, players don't know about the `AbortOnResupply` flag (nor should they care), so this just comes across as a bug.

Ultimately, `AbortOnResupply` and order queueing are just completely at odds with eachother. Either disallow order queueing for airplanes or drop `AbortOnResupply`.

pchote commented:
> Ultimately, `AbortOnResupply` and order queueing are just completely at odds with eachother. Either disallow order queueing for airplanes or drop `AbortOnResupply`.

This is at odds with the fact that all the C&C games that used `AbortOnResupply`-like logic, and several of them supported both Attack Move and queuing. To that end, I loaded up the three games I have at hand (It would be good if someone could fill in the Generals/ZH behaviour):
* RA2 supports both order queueing and attack move for vehicles, but neither for aircraft. Disappointing but not surprising considering the limitations of the engine.
* C&C3 and RA3 both RTB and cancel the current attack plus queued activities when running out of ammo during a standard attack.
* C&C3 doesn't RTB when running out of ammo during an attack move. Aircraft continue moving to the targeted location, then continue with any queued orders.
* The build of RA3 distributed through origin suffers from (what must be) a bug very similar to #17525 for only attack moves: units RTB, but then take off and perform any queued activities.

The primary action of an attack move order is to move to a location. My initial thought was that returning to base and then flying back to continue the activity was the solution here, but C&C3's solution also works.

I have pushed an update that fixes the initial set of hacks and implements an option to select between the C&C3 behaviour, and what I assume was supposed to be the RA3 behaviour (i.e. RTB during attack move and then canceling the queue). Unfortunately, this latter case requires a new set of hacks that could be dropped if we are happy to support only the C&C3 behaviour. This seems like a reasonable compromise between the different viewpoints given in #17525 and the desire to minimize code hacks here. It unfortunately makes for a much larger PR due to the `GetAttackActivity` argument changes.

tovl commented:
> C&C3 and RA3 both RTB and cancel the current attack plus queued activities when running out of ammo during a standard attack.

Are you sure about this? I already checked the behaviour in C&C3 and I found that airplanes do continue with their attack after rearming.

pchote commented:
It may depend on patch version or C&C vs KW, but yes I tested it multiple times to confirm.

tovl commented:
This was in KW. maybe what you saw has to do with losing visibility?

reaperrr commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -120,6 +122,19 @@ public override bool Tick(Actor self)
+
+				if (attackAircraft.Info.AbortOnResupply)
+				{
+					// HACK: FlyAttack may be a child activity (e.g. of AttackMoveActivity)
```
Minor general style nit: If there are more comment lines below that don't continue the sentence, please always use full stops.

reaperrr commented:
> Either disallow order queueing for airplanes or drop `AbortOnResupply`.

In addition to what @pchote said, the former would certainly cause a HUGE backlash from the player base after we just added/fixed queue support for nearly everything, while the latter would cause an uproar from modders (as it would make pre-C&C3 behavior impossible).

> (It would be good if someone could fill in the Generals/ZH behaviour)

Tested Generals ZH just now, but unless I overlooked something, the only kind of queueing it seems to have is a simple waypoint mode enabled by holding Alt, and it doesn't seem to support anything else than move commands (for both ground and air units).

On the general concept:
I feel at least part of the issue about `AbortOnResupply` is that a simple boolean isn't sufficient to cover all cases of behavior players and modders may want support for.
In my opinion this PR minus the `ResupplyDuringAttackMove`² field is good enough for prep (²once we've decided on the default behavior for RA, we should hardcode this for prep as to not introduce a field that might already be dropped again during the evolution of the bleed PR or a follow-up). 
But for Next+1 (or Next+2 if we can't decide on an approach before we branch of prep-next), I think we should consider refactoring `AbortOnResupply` more thoroughly, maybe into a `CancelOrderQueueOnResupplyOn` flag list with flags like `AttackMove`, `NormalAttack`, `ForceAttack` and so on.

pchote committed:
Fix activity cancellation on FlyAttack RTB.

* AbortOnResupply will now cancel queued activities
  in addition to the current attack.
* Resupply if no ammo is available during a standard attack.
* Don't resupply (move directly to target) if no ammo is available
  during an attack move (C&C3 style).

pchote commented:
Updated again:
 * Removed `ResupplyDuringAttackMove` and associated code hacks.
 * Implemented a proper fix to queued activity cancellation.

It turns out that the `TransformsInto*` traits all rely on canceling `NextActivity` directly, so we do need the proper fix for that rather than a local workaround here.

pchote commented:
Implementing the C&C3-style attack-move resolved your concern.

obrakmann commented on OpenRA.Game/Activities/Activity.cs:
```diff
@@ -70,12 +71,18 @@ public Activity TickOuter(Actor self)
-			if (State == ActivityState.Queued)
+			// ActivityState.Queued and firstRun are not equivalent.
+			// A queued activity that has been canceled will have ActivityState.Canceling,
+			// and we must still call OnFirstRun before Tick/OnLastRun.
```
I've been out of the loop for a bit, so I don't quite understand where the need to run a cancelled activity at all comes from. That sounds horribly broken. In the original implementation the getters would have made sure that cancelled activities weren't even visible on the queue, but I see that they've been replaced by default getters.

pchote replied:
My initial approach here was to have Cancel change `ActivityState.Queued` directly to `ActivityState.Done`, and the `return NextActivity` below first check whether it is `Queued` vs `Done`.

I had two concerns that made me change to the current approach:
* Its not as simple as just returning `null` instead of `NextActivity` - a non-queued activity may be queued after this.
* The current activity behaviour has had a lot of testing, and i'm worried that changing this behaviour may cause unexpected regressions.

If this was aimed for Next + 1 I would probably have stuck with it, but I wanted to be more conservative for the hotfix.

obrakmann replied:
> My initial approach here was to have Cancel change ActivityState.Queued directly to ActivityState.Done

Wouldn't that make the `Cancelling` state unnecessary?

> Its not as simple as just returning null instead of NextActivity - a non-queued activity may be queued after this

(You mean non-cancelled, right?)
Yes, it should be that simple*. If the getter returned null for cancelled activities, this situation could never arise. That was one of the two reasons why I did it that way - the other being to prevent cancelled activities from being run on the queue.

And if you really want to selectively only cancel the *next* activity, but not the ones following it, then that next activity should really be a child.

But I may be missing something. My impression is that OpenRA's code has become even more convoluted in the past few years, so it's likely that I don't know about edge case that should be considered.

*Unless the cancelled activity's constructor has side-effects. Which would be bug.

pchote replied:
> Wouldn't that make the `Cancelling` state unnecessary?

The `Canceling` state is still needed for activities in the `Active` state, which may require an unknown number of ticks to get themselves back into a safe configuration (e.g. moving back into the cell grid, (un)deploying, (un)docking) before completing.

> And if you really want to selectively only cancel the *next* activity, but not the ones following it, then that next activity should really be a child.

The issue comes when we want to cancel *all* activities, but have a specific reason to not cancel the current one yet (e.g. here, where we would lose target lines, or [TransformsIntoMobile](https://github.com/OpenRA/OpenRA/blob/308c64c7b1980f05bf04f98425fb8fea92d33608/OpenRA.Mods.Common/Traits/Buildings/TransformsIntoMobile.cs#L110-L115) where the undeploy needs to continue but we replace the move order(s) issued to the new actor).

Otherwise I agree with your argument here, and I don't mind changing this back if the others are happy to take the increased risk of regression.

pchote replied:
88314f8119a6bf1bae651186d91ccfd27515bc6c shows a refined version of this other approach.

pchote replied:
8c3cadb651 reworks the comment following discussion in IRC.

pchote replied:
Rebased onto the updated commit.

pchote committed:
Don't run NextActivity if it has been canceled.

tovl commented:
> It is slightly more complicated than this because we also need to suppress the target lines for the later activities.

The proper way to solve that would be to reintroduce the `IsTerminal` parameter that was cut from #16549 because we thought it wouldn't be needed anywhere. I don't think we should ever try to cancel `NextActivity` directly.

pchote commented:
> I don't think we should ever try to cancel `NextActivity` directly.

* How do we prevent this? The current situation where the engine allows it to be called but then maybe crashes later isn't good enough.
* How do we solve the `TransformsInto*` cases that also need this?

pchote commented:
It looks like this will also fix #17249, so updated the PR description.

pchote commented:
9cf3d2a647fb24e3375736a854dd5f6acd914931
7e6dcb1f90b0f97b4fa5b8c4fa47b758893a46b1
7b0a0ded48d2b0c87d6f5558f4fc3c382c069d82

https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/8024e9fa394959147c81aa97674e077c7a417cfc
